### PR TITLE
#532 customer invoice prune and validation

### LINKED ISSUE #532: Finalising a customer invoice should delete 0 quantity items
Just as it does for requisitions, stocktakes, etc., mSupply Mobile should be consistent by deleting any items left in a customer invoice with 0 quantity at the point it is finalised. This is helpful for when people add items from the master list but only actually issue a few of them.

### PR DESCRIPTION
**Pull request by Chris-Petty**
fixes #532 

Can check commit wise to avoid prettier changes. Will comment on actual changes.

Got borken somewhere in in #430 merge.

### CODE REVIEW

File: src/database/DataTypes/Transaction.js [hunk 1]
```diff
@@ -50,6 +46,14 @@ export class Transaction extends Realm.Object {
     return this.type === 'supplier_invoice';
   }
 
+  get isExternalSupplierInvoice() {
+    return this.isSupplierInvoice && this.otherParty && this.otherParty.isExternalSupplier;
+  }
+
+  get isInternalSupplierInvoice() {
```
Comment (Chris-Petty) [inline src/database/DataTypes/Transaction.js]: new getter

File: src/database/DataTypes/Transaction.js [hunk 2]
```diff
@@ -273,7 +274,10 @@ export class Transaction extends Realm.Object {
    */
   finalise(database) {
     if (this.isFinalised) throw new Error('Cannot finalise as transaction is already finalised');
+    // Should prune all invoice except internal supplier invoices
+    if (!this.isInternalSupplierInvoice) this.pruneRedundantBatchesAndItems(database);
```
Comment (Chris-Petty) [inline src/database/DataTypes/Transaction.js]: Fixes pruning on customer invoices.

File: src/pages/index.js [hunk 3]
```diff
@@ -56,6 +63,7 @@ export const FINALISABLE_PAGES = {
     finaliseText: 'finalise_supplier_invoice',
   },
   customerInvoice: {
+    checkForError: checkForCustomerInvoiceError,
```
Comment (Chris-Petty) [inline src/pages/index.js]: This + corresponding import means validation works on customer invoices finalisation again.

File: src/database/DataTypes/Transaction.js (hunk 2 above)
Comment (Chris-Petty) [reply src/database/DataTypes/Transaction.js]: @andreievg FYI, customer invoices are 'confirmed' from the get go (hence adjusting stock before being finalised), so they never run confirm() code where the pruning had been moved to.

File: src/pages/index.js (hunk 3 above)
Comment (andreievg) [reply src/pages/index.js]: As you've mentioned where the pruning was broken, the finalisation check was broken in https://github.com/sussol/mobile/pull/475/files

File: src/database/DataTypes/Transaction.js (hunk 2 above)
Comment (andreievg) [reply src/database/DataTypes/Transaction.js]: We should also not prune batches for customer invoices that have linked requisition

Comment (andreievg) [reply src/database/DataTypes/Transaction.js]: Actually rather then adding `isInternalSupplierInvoice`, we can use `isLinkedToRequisition`, no extra getter needed and covers both cases (supplier invoice and customer invoice -> is stock transfer)

File: src/database/DataTypes/Transaction.js [hunk 4]
```diff
@@ -97,14 +101,14 @@ export class Transaction extends Realm.Object {
       this.otherParty.masterLists.forEach(masterList => {
         const itemsToAdd = complement(masterList.items, this.items, item => item.itemId);
         itemsToAdd.forEach(masterListItem =>
-          createRecord(database, 'TransactionItem', this, masterListItem.item)
+          createRecord(database, 'TransactionItem', this, masterListItem.item),
```
[the remark this answers is not in the document]
Comment (Chris-Petty) [reply src/database/DataTypes/Transaction.js]: I agree, I think prettier hath derped

File: src/database/DataTypes/Transaction.js (hunk 1 above)
Comment (Chris-Petty) [reply src/database/DataTypes/Transaction.js]: Yea that's all my thinking was. Follows the scripting pattern of calling functions after they're declared too.

**Comment by Chris-Petty** [comment]
Made changes, just setting up test to check customer requisitions don't get pruned.

**Comment by Chris-Petty** [comment]
Reverted the linked approach because server prunes supplier invoices anyway when generating for receiving store.

File: src/database/DataTypes/Transaction.js [hunk 5]
```diff
@@ -162,7 +167,7 @@ export class Transaction extends Realm.Object {
    */
   addBatchIfUnique(database, transactionBatch) {
     addBatchToParent(transactionBatch, this, () =>
-      createRecord(database, 'TransactionItem', this, transactionBatch.itemBatch.item)
+      createRecord(database, 'TransactionItem', this, transactionBatch.itemBatch.item),
```
Comment (andreievg) [inline src/database/DataTypes/Transaction.js]: one more comma

File: src/database/DataTypes/Transaction.js [hunk 6]
```diff
@@ -74,6 +78,10 @@ export class Transaction extends Realm.Object {
     return this.items.length;
   }
 
+  get isLinkedToRequisition() {
+    return !!this.linkedRequisition;
```
[the remark this answers is not in the document]
Comment (Chris-Petty) [reply src/database/DataTypes/Transaction.js]: typeof should be avoided, it's a bit dirty. !!variable isn't an uncommon way to coerce a variable into a boolean. I don't even think prettier did this, I recall seeing the parentheses and thinking they're redundant. [Here's](https://www.bennadel.com/blog/1784-using-double-not-operator-for-boolean-type-casting.htm) some random ass article on it, though I actually picked this up from @interactivejunky.

[the remark this answers is not in the document]
Comment (andreievg) [reply src/database/DataTypes/Transaction.js]: Ahh ok, well if we all understand what !! means it's fine with me.

**Pull request merged**